src/library/scala/collection/mutable/TreeSet.scala

```diff
@@ -194,7 +194,7 @@ object TreeSet extends SortedIterableFactory[TreeSet] {
         new TreeSet[E](RB.fromOrderedKeys(ss.iterator, ss.size))
       case r: Range if (ordering eq Ordering.Int) || (ordering eq Ordering.Int.reverse) =>
         val it = if((ordering eq Ordering.Int) == (r.step > 0)) r.iterator else r.reverseIterator
-        new TreeSet[E](RB.fromOrderedKeys(it, r.size))
+        new TreeSet[E](RB.fromOrderedKeys(it.asInstanceOf[Iterator[E]], r.size))
```

Comment:
I'm not sure this should really be necessary. scalac unifies `E` and `Int` and it compiles. Dotty apparently goes a step further and realizes that `r` could be of type `Range & IterableOnce[T]` for some type `T <: Int` such that `E >: T` (and by extension `E >: (Int & T)` but not `E >: Int`. Since we widen the type of `r` to `Range` we lose the `T`.

But `Range` is sealed and no refinement of `T` happens in its subclasses (which are all `final`). Since the pattern match proves that we have an actual value of type `Range` it should be safe to assume that `T = Int` and therefore `E >: Int`. But maybe that's asking a bit too much of the type checker.

Reply:
@AleksanderBG opinions on this ^ ?

Reply:
I would leave the type cast for now. I have some plans to improve the situation, but they will require some finesse and I have no idea when they'll land.

While @szeiger's reasoning is correct (consider for instance `E = 5`), `scalac` is unsound in the general case - @szeiger raised the issue some time ago [here](https://github.com/lampepfl/dotty/issues/3989). The cleanest way to allow this to compile w/o a typecast would be to allow marking type argument applications as `final`, which essentially enforces that @szeiger's above reasoning applies. This would make it easier to take advantage of it in constraint inference. For clarity, `Range extends IterableOnce[final Int]` would be the equivalent of `type Range = { type T = Int }`, whereas normal application is `{ type T <: Int }`. Speculating even more, we _may_ try to infer whether some type parameter applications for `sealed` types are `final` or not, but the usual arguments for/against inferring modifiers apply. This doesn't seem like much of a priority, so again, I'm unsure when I'll get to working on it.

Reply:
Thanks, I think your `final Int` is equivalent to `=Int` in [section 4.3 of @Blaisorblade's paper](http://lampwww.epfl.ch/~hmiller/scala2013/resources/pdfs/paper5.pdf), but in general if we could infer that from the sealed-ness of the type that'd be nicer (changing the children of a sealed trait in any way already breaks the API of the sealed trait, so inferring something here is fine I think).

Reply:
It's definitely similar, but I think they differ in some details (I believe it's not actually necessary to forbid inheritance the way Paolo's paper suggests, for instance). We/I won't know for certain until I write it down and work everything out.

As to inferring `final/=` - the problem I'm seeing is unexpected breakages. If it is inferred, then adding a new class extending, say, `Range` can silently remove the property and break code in unexpected ways. Needing to handle a new subclass is one thing, making already existing code not typecheck because an invisible property was previously inferred and now for potentially unclear reasons is not is another. Consider that we don't infer variance either, even though we could.

Reply:
I think the breakage potential is limited compared to how useful it would be.